### PR TITLE
[Test][XPU] Make test/export/{test_export,test_draft_export}.py device-agnostic and enable on XPU

### PR DESCRIPTION
**Pull request by tszulist-hbn**
## Summary

This PR makes selected export tests device-agnostic and enables their accelerator coverage on XPU.

## What changed

- Added device-templated accelerator test classes for `test_export.py` and `test_draft_export.py`.
- Moved 9 CUDA-gated tests to accelerator-generic coverage so they run on CUDA and XPU.
- Removed an unnecessary Triton requirement from tensor metadata validation.
- Generalized the scaled dot product attention and accelerator memory usage tests.
- Added hardware classification metadata to the test classes.
- Test-only change; no CPU coverage was added or removed.

## Test coverage delta

| Test suite | Before | After |
|---|---:|---:|
| `test_export.py`, XPU | 509 passed / 12 skipped / 2 xfailed | 517 passed / 11 skipped / 2 xfailed |
| `test_draft_export.py`, XPU | 23 passed / 2 skipped | 25 passed / 2 skipped |
| Derived export suites, XPU | — | 4,239 passed / 31 skipped / 71 xfailed |

No unexpected failures occurred. The derived suites collect 54 fewer test instances because tests moved into device-templated classes are no longer redundantly re-instantiated by those suites.

### CODE REVIEW

**Comment by EikanWang** [comment]
The PR should be a draft PR first if it is xpu-specific. Meanwhile, this PR is not specific to xpu.

**Comment by tszulist-hbn** [comment]
My bad, converted to draft.

File: test/export/test_draft_export.py [hunk 1]
```diff
@@ -285,7 +296,7 @@ def forward(self, a, b):
                 ):
                     torch.ops.mylib.foo8(*inp)
 
-    @unittest.skipIf(not torch.cuda.is_available(), "Requires cuda")
+    @requires_accelerator
     def test_missing_meta_kernel_guard(self):
```
Comment (pbielak) [inline test/export/test_draft_export.py]: Why not extract to separate class `TestDraftExportDevice` and use `device` parameter with `@onlyAccelerator`?

File: test/export/test_draft_export.py [hunk 2]
```diff
@@ -768,11 +779,11 @@ def forward(self, x, y):
             )
 
     @unittest.skipIf(
-        not torch.cuda.is_available()
-        or torch.cuda.get_device_properties(0).total_memory < 2**28,
+        not torch.accelerator.is_available()
```
Comment (pbielak) [inline test/export/test_draft_export.py]: Ditto + use `@onlyAccelerator` + use `largeTensorTest`

File: test/export/test_export.py [hunk 3]
```diff
@@ -2628,7 +2634,7 @@ def forward(self, x, y, out):
         # For non-functional graph module, out_copy is not mutated
         self.assertEqual(out_copy2, out_copy3)
 
-    @requires_cuda_and_triton
+    @requires_gpu
```
Comment (pbielak) [inline test/export/test_export.py]: Why this change? Triton is still required here

File: test/export/test_export.py [hunk 4]
```diff
@@ -2639,8 +2645,8 @@ def forward(self, x, y):
                 return out
 
         args = (
-            torch.randn(3, device="cuda"),
-            torch.randn(3, device="cuda"),
+            torch.randn(3, device=GPU_TYPE),
```
Comment (pbielak) [inline test/export/test_export.py:2651]: Extract to separate class `TestExportDevice` and use `device` parameter

File: test/export/test_export.py [hunk 5]
```diff
@@ -10922,10 +10928,10 @@ def forward(self, x):
                 len([node for node in gm.graph.nodes if node.op == "placeholder"]), 1
             )
 
-    @requires_cuda_and_triton
+    @requires_gpu
     @testing.expectedFailureCppRuntime
     def test_export_associative_scan_symbol_dim(self):
```
Comment (pbielak) [inline test/export/test_export.py:10936]: Ditto

File: test/export/test_export.py [hunk 6]
```diff
@@ -10947,10 +10953,10 @@ def forward(self, x):
         module_out = Foo()(xs)
         self.assertTrue(torch.allclose(ep.module()(xs), module_out))
 
-    @requires_cuda_and_triton
+    @requires_gpu
     @testing.expectedFailureCppRuntime
     def test_export_associative_scan_symbol_scandim(self):
-        device = torch.device("cuda")
+        device = torch.device(GPU_TYPE)
```
Comment (pbielak) [inline test/export/test_export.py:10962]: Ditto

File: test/export/test_export.py [hunk 7]
```diff
@@ -10972,12 +10978,12 @@ def forward(self, x):
...
         self.assertTrue(torch.allclose(ep.module()(xs), module_out))
 
-    @requires_cuda_and_triton
+    @requires_gpu
     def test_export_associative_scan_lifted_buffers(self):
         if "cpp_runtime_nonstrict" in self.id():
             self.skipTest("TODO Unexpected success in OSS but not in fbcode.")
 
-        device = torch.device("cuda")
+        device = torch.device(GPU_TYPE)
```
Comment (pbielak) [inline test/export/test_export.py:10989]: Ditto

File: test/export/test_export.py [hunk 8]
```diff
@@ -17870,7 +17876,8 @@ def forward(self, arg0_1: "f32[2, 4]", arg1_1: "f32[4]"):
             ignore_empty_lines=True,
         )
 
-    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @requires_accelerator
+    @unittest.skipIf(TEST_MPS, "UntypedStorage.resize_ is not implemented for MPS")
```
Comment (pbielak) [inline test/export/test_export.py]: How did you test this?

File: test/export/test_export.py [hunk 9]
```diff
@@ -17870,7 +17876,8 @@ def forward(self, arg0_1: "f32[2, 4]", arg1_1: "f32[4]"):
...
     def test_module_to_with_shared_weights(self):
```
Comment (pbielak) [inline test/export/test_export.py]: Also export to separate class

File: test/export/test_export.py [hunk 10]
```diff
@@ -18038,7 +18045,7 @@ def fn(x):
         self.assertEqual(x.sin(), ep.module()(x))
         pytree._deregister_pytree_node(torch.FunctionSchema)
 
-    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @requires_accelerator
```
Comment (pbielak) [inline test/export/test_export.py:17977]: Ditto

File: test/export/test_export.py [hunk 11]
```diff
@@ -19154,7 +19168,7 @@ def forward(self, x):
             len(list(new_ep.graph.nodes)[-1].args[0]), len(signature.output_specs)
         )
 
-    @requires_cuda_and_triton
+    @requires_accelerator
```
Comment (pbielak) [inline test/export/test_export.py]: Ditto

**Review by pbielak — CHANGES_REQUESTED** [review]

File: test/export/test_draft_export.py (hunk 1 above)
Comment (tszulist-hbn) [reply test/export/test_draft_export.py]: Done. That also let me drop the module-level device_type global from this file entirely.

File: test/export/test_draft_export.py (hunk 2 above)
Comment (tszulist-hbn) [reply test/export/test_draft_export.py]: Done. `largeTensorTest` is slightly stricter than what was there before: it gates on free memory (mem_get_info) rather than total device memory, so on a busy device the test now skips instead of OOMing. I kept 2**28 and the intent comment ("16 MB is enough to run the test; ask for more so a regression OOMs instead of silently passing").

File: test/export/test_export.py (hunk 8 above)
Comment (tszulist-hbn) [reply test/export/test_export.py]: Dropped this.

File: test/export/test_export.py (hunk 3 above)
Comment (tszulist-hbn) [reply test/export/test_export.py]: To be clear the original was not a relaxation: `requires_gpu` is `skipUnless(HAS_GPU) and HAS_GPU = HAS_CUDA_AND_TRITON or HAS_XPU_AND_TRITON or HAS_MTIA_AND_TRITON` (`inductor_utils.py:82`), the same disjunction. I picked `requires_gpu` because it was already imported in the file.
Anyway, since it may be confusing, so replaced it with explicit `requires_gpu_and_triton`.

File: test/export/test_export.py (hunk 5 above)
Comment (tszulist-hbn) [reply test/export/test_export.py:10936]: As above. Replaced.

File: test/export/test_export.py (hunk 11 above)
Comment (tszulist-hbn) [reply test/export/test_export.py]: Replaced.

File: test/export/test_export.py (hunk 9 above)
Comment (tszulist-hbn) [reply test/export/test_export.py]: Dropped TEST_MPS

**Review by sylvesterkaczmarek — COMMENTED** [review]
The current structure is much cleaner than hard-coded CUDA branches: accelerator-specific export cases now own a `device` parameter, memory accounting uses `torch.accelerator`, and Triton cases use `GPU_TYPE`. That keeps the test intent while making backend coverage explicit instead of inferred from string checks.

File: test/export/test_export.py [hunk 12]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
...
+            with self.assertRaisesRegex(RuntimeError, "Couldn't swap Embedding.weight"):
+                _ = torch.export.export(
+                    BarBar(),
+                    (),
+                    {"x": torch.randn(4, 4, 4, device=device_type)},
+                    strict=False,
+                ).module()
+
+    @onlyAccelerator
+    @requires_gpu_and_triton
```
Comment (pbielak) [inline test/export/test_export.py]: Can be replaced with `requires_triton` (GPU is already covered by `instantiate_device_type_tests` and `onlyAccelerator`

File: test/export/test_export.py [hunk 13]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
...
+            torch.randn(3, device=device),
+        )
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Raw Triton kernel calls are not supported by non-strict torch.export",
+        ):
+            export(M(), args, strict=False)
+
+    @onlyAccelerator
+    @requires_gpu_and_triton
```
Comment (pbielak) [inline test/export/test_export.py]: Ditto

File: test/export/test_export.py [hunk 14]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
...
+                return associative_scan(
+                    self.combine_fn, x, 2, combine_mode=combine_mode
+                )
+
+        ep = export(Foo(), (xs,), dynamic_shapes={"x": {1: dim1}})
+        module_out = Foo()(xs)
+        self.assertTrue(torch.allclose(ep.module()(xs), module_out))
+
+    @onlyAccelerator
+    @requires_gpu_and_triton
```
Comment (pbielak) [inline test/export/test_export.py]: Ditto

File: test/export/test_export.py [hunk 15]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
...
+                return associative_scan(
+                    self.combine_fn, x, 1, combine_mode=combine_mode
+                )
+
+        ep = export(Foo(), (xs,), dynamic_shapes={"x": {1: dim1}})
+        module_out = Foo()(xs)
+        self.assertTrue(torch.allclose(ep.module()(xs), module_out))
+
+    @onlyAccelerator
+    @requires_gpu_and_triton
```
Comment (pbielak) [inline test/export/test_export.py]: Ditto

**Review by pbielak — CHANGES_REQUESTED** [review]

**Review by pbielak — APPROVED** [review]

File: test/export/test_export.py [hunk 16]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
             self.assertEqual(eager_out[1], decomp_out[1])
 
 
+@unittest.skipIf(not torchdynamo.is_dynamo_supported(), "dynamo isn't support")
+class TestExportDevice(TorchTestCase):
```
Comment (etaf) [inline test/export/test_export.py]: The derived suites (test_serdes, test_retraceability, test_export_strict, test_strict_export_v2, test_export_training_ir_to_run_decomp, test_nativert, test_cpp_serdes) only wrap `test_export.TestExport`, and they get their coverage by swapping out the module-level `export`. So anything moved into this class no longer gets the serdes/retrace/strict/training-IR variants.

For `test_module_to_with_shared_weights` and `test_exception` that costs nothing, because they call `torch.export.export` directly. The same goes for the raw Triton test, which raises before any round-trip. The three `test_export_associative_scan_*` tests are different: they call `export(...)`, so their derived variants really do exercise serialization and retracing of `associative_scan` with a symbolic dim and with lifted buffers. The only associative_scan test left in `TestExport` is `test_export_associative_scan_pointwise_cpu`, which covers neither case. The description calls the 54 dropped instances redundant. For these three tests they aren't: that's a real coverage loss.

A side effect: `@testing.expectedFailureCppRuntime` at L19265/L19286 and the `"cpp_runtime_nonstrict" in self.id()` skip at L19308 are now dead code. Nothing ever gives these tests a cpp_runtime suffix anymore.

I know pbielak asked for these to be extracted. Still, I'd keep these three in `TestExport` using the pattern the file already uses for GPU tests (`@requires_gpu` plus `device=GPU_TYPE`, e.g. around L2488-L2626). That still runs them on CUDA and XPU, and the decorators keep working. The other option is to add the per-device classes to each derived suite's `tests` list, but that's more work for the same result.

File: test/export/test_export.py [hunk 17]
```diff
@@ -18632,6 +18409,8 @@ def forward(self, x):
 
 @unittest.skipIf(not torchdynamo.is_dynamo_supported(), "dynamo isn't support")
 class TestOneOffModelExportResult(TestCase):
+    hw_classification = HardwareClassification.GENERIC
```
Comment (etaf) [inline test/export/test_export.py]: This class contains `test_scaled_dot_product_attention_gpu`, which this PR just enabled on XPU. `TestExport` (L589) also has many `@requires_gpu` tests. By the `HardwareClassification` docstring, `GENERIC` means "does not depend on a particular accelerator". So `--hw-classification ACCELERATOR` filters these GPU tests out, and a `GENERIC`-only run on a CPU box just skips them. With either split they never run.

For the SDPA test, the simple fix is to move it into `TestExportDevice` with `@onlyAccelerator` and use the `device` argument instead of the module-level `device_type`. For `TestExport`, tagging the whole class `GENERIC` misdescribes it. I'm not sure what the intended convention is for mixed classes, so it may be better to leave it untagged until that's settled.

File: test/export/test_export.py [hunk 18]
```diff
@@ -19319,8 +19087,272 @@ def forward(self, a, b, c, d):
...
+                with torch.amp.autocast(device_type=device_type):
+                    y = self.mod(x)
+                return y
+
+        with torch.no_grad():
+            with self.assertRaisesRegex(RuntimeError, "Couldn't swap Embedding.weight"):
+                _ = torch.export.export(
+                    BarBar(),
+                    (),
+                    {"x": torch.randn(4, 4, 4, device=device_type)},
```
Comment (etaf) [inline test/export/test_export.py]: This should be `device=device`, like the other tests in this class, not `device_type`.

**Review by etaf — CHANGES_REQUESTED** [review]
